gamemode/modules/main_menu/cl_mainmenu_create.lua

```diff
@@ -0,0 +1,338 @@
+    end
+
+    function textEntry:OnTextChanged()
+        local curText = self:GetText()
+        if curText and #curText > maxLength then
+            self:SetText( string.Left( curText, maxLength ) )
+            self:SetCaretPos( maxLength )
+            surface.PlaySound( "resource/warning.wav" )
+        end
+
+        data.action( curText )
+    end
+
+    data.action( "" )
+
+    return textEntry
+end
+
+function mainMenu.createColorEntry( data )
+    data.entryType = "DColorMixer"
+    local picker = mainMenu.createEntry( data )
+    picker:SetColor( data.defaultColor )
+    picker:SetAlphaBar( false )
+    picker:SetPalette( false )
+    picker:SetWangs( false )
+
+    function picker:Paint( _w, _h )
+        DisableClipping( true )
+
+        draw.RoundedBox( 6, -10, -10, _w + 20, _h + 20, Color( 158, 122, 74 ) )
+
+        DisableClipping( false )
+    end
+
+    function picker:ValueChanged( col )
+        data.action( col )
+    end
+
+    local preview = vgui.Create( "DPanel", mainMenu.container )
+    preview:SetSize( 30, 30 )
+
+    function preview:PerformLayout()
+        local x, y = data.label:GetPos()
+        local _w, _h = data.label:GetSize()
+        self:SetPos( x + _w + 10, y + _h * 0.5 - 13 )
+    end
+
+    function preview:Paint( _w, _h )
+        draw.RoundedBox( 10, 3, 3, _w - 6, _h - 6, picker:GetColor() )
+    end
+
+    data.action( data.defaultColor )
+end
+
+function mainMenu.createRadioMenuEntry( data )
+    data.entryType = "DPanel"
+    local panel = mainMenu.createEntry( data )
+    panel.Paint = nil
+
+    local spacing = 15
+
+    local btns = {}
+    for k, option in pairs( data.options ) do
+        local last = k == #data.options
+
+        local btn = vgui.Create( "DButton", panel )
+        btn:SetFont( "RVR_StartMenuButton" )
+        btn:SetText( option )
+        btn:Dock( TOP )
+        btn:DockMargin( 0, 0, 0, last and 0 or spacing )
+        btn.option = option
+
+        function btn:PerformLayout()
+            local _h = panel:GetTall() + spacing
+            _h = _h / #data.options - spacing
+            self:SetTall( _h )
+        end
+
+        function btn:DoClick()
+            for _, other in pairs( btns ) do
+                if other ~= self then
+                    other:SetSelected( false )
+                end
+            end
+
+            self:SetSelected( true )
+        end
+
+        function btn:SetSelected( v )
+            self.selected = v
+            self:SetTextColor( v and mainMenu.darkerYellow or mainMenu.darkBrown )
+
+            if v then
+                data.action( self.option )
+            end
+        end
+
+        function btn:Paint( _w, _h )
+            local bgCol = self.selected and mainMenu.darkBrown or mainMenu.darkerYellow
+            draw.RoundedBox( 10, 0, 0, _w, _h, bgCol )
+        end
+
+        btn:SetSelected( k == 1 )
+        table.insert( btns, btn )
+    end
+end
+
+local joinModeLookup = {
+    ["PUBLIC"] = 0,
```

Comment:
would defining  the table like `PUBLIC = 0` be better

Reply:
no